File: p23/p23.java

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}
 
public class p23 {
    public static void heapAdjust(int cur,int max,int[] h,ListNode[] n)
    {
        int i=cur;
        while((i<<1)<=max)
        {
            int p=i<<1;

            int a=n[h[i]].val<=n[h[p]].val ? i:p;
            if(p<max) a=n[h[a]].val<=n[h[p+1]].val ? a:p+1;

            if(a==i) return;

            int tmp=h[i]; h[i]=h[a]; h[a]=tmp;
            i=a;
        }
    }
    public static ListNode mergeKLists(ListNode[] lists) {
        ListNode ans=null,p=null;
        int k=lists.length;
        int[] h = new int[k+1]; /* min heap */
        ListNode[] n = new ListNode[k+1];

        int idx=0;
        for(int i=0;i<k;i++)
            if(lists[i]!=null)
            {
                idx++;
                h[idx]=idx;
                n[idx]=lists[i];
            } 
        k=idx;

        for(int i=k/2;i>=1;i--)
            heapAdjust(i,k,h,n);

        while(k>0)
        {
            int tmp=h[1];
        	if(ans==null)
        		ans=p=n[tmp];
        	else
        	{
        		p.next=n[tmp];
        		p=p.next;
        	}
        	n[tmp]=n[tmp].next;
            if(n[tmp]==null)
            {
                k--;
                if(k==0) break;
                h[1]=h[k+1];
            }

            heapAdjust(1,k,h,n);
        }
        return ans;
    }
    public static void main(String[] args) {
        ListNode[] lists = new ListNode[3];
        lists[0]=new ListNode(2);
        lists[0].next=new ListNode(6);
        lists[0].next.next=new ListNode(10);
        lists[1]=new ListNode(5);

        ListNode res=mergeKLists(lists);
        while(res!=null)
        {
            System.out.println("val:"+res.val); res=res.next;
        } 
    }
}

File: p23/p23.cpp
// Implemented a min heap by myself.
// O(K) extra space. O(N*logK) time. N is the number of all numbers.

class PriorityQueue
{
public:
    vector<ListNode*> heap;

    PriorityQueue() {
        heap = std::vector<ListNode*>(1);
    }

    void push(ListNode* p);
    ListNode* pop();
};

void PriorityQueue::push(ListNode* p) {
    if (p == NULL)
        return;

    heap.push_back(p);

    int j = heap.size()-1;
    int i = j>>1;
    
    while (i >= 1) {
        if (heap[i]->val <= heap[j]->val)
            break;

        ListNode *tmp = heap[i];
        heap[i] = heap[j];
        heap[j] = tmp;

        j = i;
        i = j>>1;
    }
}

ListNode* PriorityQueue::pop() {
    ListNode *res = heap[1];

    heap[1] = heap.back();
    heap.pop_back();

    int i = 1, j = 2;
    while (j < heap.size()) {
        int k = i;
        if (heap[j]->val < heap[k]->val) {
            k = j;
        }

        if (j+1 < heap.size() && heap[j+1]->val < heap[k]->val) {
            k = j+1;
        }

        if (k == i)
            break;

        ListNode *tmp = heap[i];
        heap[i] = heap[k];
        heap[k] = tmp;

        i = k;
        j = i<<1;
    }

    return res;
}

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        PriorityQueue pq;

        for (int i = 0; i < lists.size(); ++i)
            if (lists[i] != NULL)
                pq.push(lists[i]);
        
        ListNode* resHead = new ListNode(0);
        ListNode* cur = resHead;

        while (pq.heap.size()>1) {
            cur->next = pq.pop();
            pq.push(cur->next->next);

            cur = cur->next;
        }

        return resHead->next;
    }
};
